File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_py/code_output/NLPDataProcessor2.py
import re
from typing import List, Dict, Tuple

class NLPDataProcessor2:
    class WordFrequency:
        def __init__(self, word: str, frequency: int):
            self.word = word
            self.frequency = frequency

        def get_word(self) -> str:
            return self.word

        def get_frequency(self) -> int:
            return self.frequency

        def __repr__(self) -> str:
            return f"WordFrequency{{word='{self.word}', frequency={self.frequency}}}"

        def __eq__(self, other) -> bool:
            if not isinstance(other, NLPDataProcessor2.WordFrequency):
                return False
            return self.word == other.word and self.frequency == other.frequency

        def __hash__(self) -> int:
            return hash((self.word, self.frequency))

        @staticmethod
        def by_frequency_then_word():
            # Returns a key function for sorting: descending frequency, then ascending word.
            return lambda wf: (-wf.frequency, wf.word)

    def process_data(self, string_list: List[str]) -> List[List[str]]:
        words_list: List[List[str]] = []
        pattern = re.compile(r'[^a-zA-Z\s]')
        for s in string_list:
            processed = pattern.sub('', s.lower())
            if not processed:
                words_list.append([])
            else:
                words_list.append(processed.split())
        return words_list

    def calculate_word_frequency(self, words_list: List[List[str]]) -> List[WordFrequency]:
        frequency_map: Dict[str, int] = {}
        order_map: Dict[str, int] = {}
        index = 0

        for words in words_list:
            for word in words:
                if word not in frequency_map:
                    order_map[word] = index
                    index += 1
                frequency_map[word] = frequency_map.get(word, 0) + 1

        word_frequencies: List[NLPDataProcessor2.WordFrequency] = []
        for word, freq in frequency_map.items():
            if freq > 1 or word == '%%%':
                word_frequencies.append(NLPDataProcessor2.WordFrequency(word, freq))

        # Sort: descending frequency, then ascending order of first appearance.
        word_frequencies.sort(key=lambda wf: (-wf.frequency, order_map[wf.word]))
        return word_frequencies

    def process(self, string_list: List[str]) -> List[WordFrequency]:
        words_list = self.process_data(string_list)
        return self.calculate_word_frequency(words_list)

File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_cpp/code_output/NLPDataProcessor2.cpp
#include <iostream>
#include <vector>
#include <string>
#include <regex>
#include <sstream>
#include <algorithm>
#include <map>
#include <cctype>

class NLPDataProcessor2 {
public:
    struct WordFrequency {
        std::string word;
        int frequency;

        WordFrequency(const std::string& word, int frequency)
            : word(word), frequency(frequency) {}

        std::string getWord() const { return word; }
        int getFrequency() const { return frequency; }

        bool operator==(const WordFrequency& other) const {
            return word == other.word && frequency == other.frequency;
        }

        std::string toString() const {
            return "WordFrequency{word='" + word + "', frequency=" + std::to_string(frequency) + "}";
        }

        static bool byFrequencyThenWord(const WordFrequency& a, const WordFrequency& b) {
            if (a.frequency != b.frequency)
                return a.frequency > b.frequency;
            return a.word < b.word;
        }
    };

    std::vector<std::vector<std::string>> processData(const std::vector<std::string>& stringList) {
        std::vector<std::vector<std::string>> wordsList;
        std::regex nonAlphaWhitespace("[^a-zA-Z\\s]");

        for (const auto& str : stringList) {
            std::string lower = str;
            std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);
            std::string processed = std::regex_replace(lower, nonAlphaWhitespace, "");

            std::vector<std::string> words;
            if (!processed.empty()) {
                std::istringstream stream(processed);
                std::string token;
                while (stream >> token) {
                    words.push_back(token);
                }
            }
            wordsList.push_back(words);
        }
        return wordsList;
    }

    std::vector<WordFrequency> calculateWordFrequency(const std::vector<std::vector<std::string>>& wordsList) {
        std::map<std::string, int> frequencyMap;
        std::vector<std::string> order;
        std::map<std::string, int> orderMap;

        for (const auto& words : wordsList) {
            for (const auto& word : words) {
                if (frequencyMap.find(word) == frequencyMap.end()) {
                    order.push_back(word);
                    orderMap[word] = order.size() - 1;
                }
                frequencyMap[word]++;
            }
        }

        std::vector<WordFrequency> wordFrequencies;
        for (const auto& entry : frequencyMap) {
            const std::string& word = entry.first;
            int frequency = entry.second;
            if (frequency > 1 || word == "%%%") {
                wordFrequencies.emplace_back(word, frequency);
            }
        }

        std::sort(wordFrequencies.begin(), wordFrequencies.end(),
            [&orderMap](const WordFrequency& a, const WordFrequency& b) {
                if (a.frequency != b.frequency)
                    return a.frequency > b.frequency;
                return orderMap[a.word] < orderMap[b.word];
            });

        return wordFrequencies;
    }

    std::vector<WordFrequency> process(const std::vector<std::string>& stringList) {
        auto wordsList = processData(stringList);
        return calculateWordFrequency(wordsList);
    }
};